libs/core/langchain_core/runnables/rate_limiter.py

```diff
@@ -0,0 +1,210 @@
+"""Provides a rate limiter based on a token bucket algorithm.
+
+This rate limiter is designed to only limit based on the number of requests
+per second.
+
+It does not take into account the size of the request or any other factors.
+"""
+from __future__ import annotations
+
+import abc
+import asyncio
+import threading
+import time
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Optional,
+    TypeVar,
+)
+
+from langchain_core.runnables.base import (
+    Input,
+    Output,
+    Runnable,
+    RunnableBindingBase,
+    RunnableLambda,
+)
+
+if TYPE_CHECKING:
+    from langchain_core.callbacks.manager import (
+        AsyncCallbackManagerForChainRun,
+        CallbackManagerForChainRun,
+    )
+
+    T = TypeVar("T", CallbackManagerForChainRun, AsyncCallbackManagerForChainRun)
+U = TypeVar("U")
+
+
+class BaseRateLimiter(abc.ABC):
+    """Base class for rate limiters.
+
+    Usage of the base limiter is through the wait and await_ methods depending
+    on whether running in a sync or async context.
+    """
+
+    @abc.abstractmethod
+    def _consume(self) -> bool:
+        """Consume the given amount of tokens if possible.
+
+        Returns:
+            True means that the tokens were consumed, and the caller can proceed to
+            make the request. A False means that the tokens were not consumed, and
+            the caller should try again later.
+        """
+
+    @abc.abstractmethod
+    async def _aconsume(self) -> bool:
```

Comment:
does this need to be in the base class? or is this specific to token counting rate limiting methods?

Reply:
Base class useful if we want to support rate limit between different processes, could be overkill since it'll probably cause serialization complications